proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -24,77 +24,150 @@ vector/matrix operations described in [0029].
 ## Motivation
 
-See [0029] for general background around the need for these new operations.
+Modern GPUs have dedicated silicon to accelerate matrix operations, but HLSL
+doesn't provide a mechanism to easily utilize these units. Evaluation of
+matrix-vector operations (multiply, muladd, accumulation) in HLSL was previously
+scalarized at the DXIL level making it hard to employ these specialized units.
+This proposal builds on the "Long vectors" feature described in [0026],
+providing a mechanism to express matrix-vector ops in HLSL that can be lowered
+to the DXIL ops described [0029], these primitives provide the right level of
+abstraction for hardware acceleration.
 
 An HLSL API needs to be defined to expose these new operations in a way that:
-* work well with existing HLSL APIs
+* works well with existing HLSL APIs
 * is expected to work well with future HLSL APIs in the same problem space
 * can be implemented reasonably in DXC and cleanly in clang
 
-This design builds on the "long vectors" feature described in [0026].
-
 [0026]: 0026-hlsl-long-vector-type.md
 
 ## Proposed solution
 
-First strawman:
+This API will be implemented using HLSL code.  The exact mechanism for getting
+this code into a developer's shader is TBD, but implementations have a few
+possible options, including:
+
+* Developers must explicitly #include a header file
+* The compiler force-includes the header file
+* The compiler force-includes a precompiled version of the header file
+
+The header-implementation accesses the DXIL operations described in [0029] by
+calling low-level builtins. These builtins should be considered implementation
+details and users should not call them directly. However, since they are a part
+of the implementation plan for the first implementation of this API they are
+described [below](#builtins).
+
+Since this API is currently only supported by DirectX, all the new types /
+methods described in it are placed in the `dx` namespace. Within this namespace,
+a `linalg` namespace is also added to group together types and methods related
+to linear algebra.
+
+Throughout this API, template parameters are used to store values that must be
+known at compile time, while member variables or function arguments are used to
+store values that may only be determined at runtime.
+
+This API defines the following supporting types:
+
+* `struct dx::linalg::MatrixRef`
+  * Reference to a matrix stored in a ByteAddressBuffer.   
+* `struct dx::linalg::RWMatrixRef`
+  * Reference to a matrix stored in a RWByteAddressBuffer.
+* `struct dx::linalg::VectorRef`
+  * Reference to a vector stored in a ByteAddressBuffer.
+* `struct dx::linalg::RWVectorRef`
+  * Reference to a vector stored in a RWByteAddressBuffer.
+* `struct dx::linalg::Vector`
+  * Wrapper around a vector, allowing the elements of the vector to be
+    reinterpreted in various ways.
+* `enum dx::linalg::DataType`
+  * Enum describing various data types that can be used to applied to matrices
```

Comment:
Did you mean to say something more like 'that can be applied to elements of matrices and vectors'?